File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/decorator/lab/Barista.java
// Barista.java

import java.util.Vector;

public class Barista
{
	public Coffee makeCoffee(Vector<String> aDescription)
	{
		switch (aDescription.size())
		{
			case 1:
				if (aDescription.elementAt(0).equals("coffee"))
				{
					return new CoffeeWithoutMilkWithoutSugar();
				}
				else
				{
					// throw an exception.
				}
				break;
			case 2:
				if (aDescription.elementAt(0).equals("coffee"))
				{
					if (aDescription.elementAt(1).equals("milk"))
					{
						return new CoffeeWithMilkWithoutSugar();
					}
					else if (aDescription.elementAt(1).equals("sugar"))
					{
						return new CoffeeWithoutMilkWithSugar();
					}
					else
					{
						// throw an exception.
					}
				}
				else
				{
					// throw an exception.
				}
				break;
			case 3:
				if (aDescription.elementAt(0).equals("coffee"))
				{
					if (aDescription.elementAt(1).equals("milk"))
					{
						if (aDescription.elementAt(2).equals("sugar"))
						{
							return new CoffeeWithMilkWithSugar();
						}
						else
						{
							// throw an exception.
						}
					}
					else if (aDescription.elementAt(1).equals("sugar"))
					{
						if (aDescription.elementAt(2).equals("milk"))
						{
							return new CoffeeWithMilkWithSugar();
						}
						else
						{
							// throw an exception.
						}
					}
					else
					{
						// throw an exception.
					}
				}
				else
				{
					// throw an exception.
				}
				break;
			default:
				// throw an exception.
				break;
		}
		return null; // when we add exceptions, we can remove this.
	}
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/decorator/solution/Barista.m
% Barista.m

classdef Barista < handle
	methods
		function result = makeBeverage(theBarista, aDescription)
			result = []; % when we add exceptions, we can remove this.

			if (aDescription.getSize == 0)
				% throw an exception.
			end
	
			if ~strcmp(aDescription.get(1), 'coffee')
				% throw an exception.
			end
	
			result = Coffee;
	
			for i = 2:aDescription.getSize
				if strcmp(aDescription.get(i), 'milk')
					result = WithMilk(result);
				elseif strcmp(aDescription.get(i), 'sugar')
					result = WithSugar(result);
				else
					% throw an exception.
				end
			end
		end
	end
end
